### PR TITLE
Add Enum.none?/1 and Enum.none?/2

### PR DESCRIPTION
**Pull request by endersstocker**

### CODE REVIEW

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -1390,6 +1390,38 @@ defmodule Enum do
...
+      false
+
+  """
+  @spec none?(t) :: boolean
+  @spec none?(t, (element -> as_boolean(term))) :: boolean
+
+  def none?(enumerable, fun \\ fn(x) -> x end)
+
+  def none?(enumerable, fun) do
+    !any?(enumerable, fun)
```
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: `not any?(enumerable, fun)`
will be more readable in this case

**Comment by endersstocker** [comment]
@eksperimental Thanks!

**Comment by lexmag** [comment]
I think there was already discussion about adding these.
"It doesn't bring much value to have it in stdlib" – was a conclusion if I'm correct (and I actually think so). :bowtie:

**Comment by eksperimental** [comment]
well, i think having this one makes sense.
as it is the opossite of Enum.any?/2

**Comment by josevalim** [comment]
Yes, let's bring those discussions to the elixir-lang-core mailing list. It is not worth much exactly because it is the same as `not any?` and `one?` can be easily implemented with `find`. Thank you!

**Comment by endersstocker** [comment]
@josevalim `Enum.find` will match when multiple invocations are truthy, as with `Enum.any?`. `Enum.one?`, on the other hand, matches when exactly one invocation is truthy.

**Comment by josevalim** [comment]
@endersstocker you can still do a filter and check you got a single element back. :)

**Comment by endersstocker** [comment]
@josevalim Yes, but that would require traversing the entire list. `Enum.one?` short circuits as soon as a second truthy invocation is found.

**Comment by endersstocker** [comment]
An example use case: I have a list of `technician` data structures. I need to ensure that exactly one of those technicians is the job foreman.

**Comment by lexmag** [comment]
@endersstocker there is a way to make `Enum.one?` in one line:
```elixir
match? [_], Enum.filter(enum, fun)
```
and short-circuit:
```elixir
match? [_], Stream.filter(enum, fun) |> Enum.take(2)
```

**Comment by endersstocker** [comment]
@lexmag That is definitely a better implementation. I’ve updated my branch to utilize it.

I based `Enum.one?` on Ruby’s `Enumerable#one?`. Even though it’s doable in single line in Elixir, that doesn’t mean it’s intuitive enough to be a pattern (as opposed to a library function). `Enum.one?` vastly simplifies the solution to a common enough problem (detecting uniqueness). `[1, 2, 3]  |> Enum.sum` is an improvement over `[1, 2, 3]  |> Enum.reduce(0, &+/2)`, but only marginally in comparison.

**Comment by eksperimental** [comment]
Sorry, if my comment is absolutely off..
but wouldn't 
`Enum.one?([1, 2, 3], fn(x) -> rem(x, 2) == 0 end)`
be the same as doing
`Enum.count([1, 2, 3], fn(x) -> rem(x, 2) == 0 end) == 1`

**Comment by endersstocker** [comment]
@eksperimental The return values would be the same, but `Enum.count` would traverse the entire enumerable. `Enum.one?`, on the other hand would short circuit as soon as a second truthy invocation was discovered.

**Comment by eksperimental** [comment]
Interesting. I would be more inclined to a more generic function. such as `Enum.exactly(enumerable, fun, count)`

**Comment by lexmag** [comment]
Soo big discussion here, that's why we need it in elixir-lang-core mailing list at first place.

**Comment by lexmag** [comment]
`Enum.exactly(enumerable, fun, count)` seems reasonable and seems covers `none?`, `one?` at once. I'm wondering what other languages have for it?

**Comment by endersstocker** [comment]
@eksperimental Singularity is a meaningful concept in-and-of-itself. For instance, we wouldn’t say that `Enum.uniq` demands `Enum.with_n_allowed_instances`, just because uniqueness can be abstracted to be a case of a certain number of allowed instances.

Additionally I can think of many more scenarios where a programmer would need to determine that criteria is met uniquely (one time) as opposed to a given number of times.

**Comment by eksperimental** [comment]
@endersstocker I see your point, and it's a valid one.
I think we are leavinga lot of people out of the discussion because not everyone from mailing list is here, and I'm sure they will have valuable opinions to add.
@endersstocker could you please send the proposal there and a reference to this talk.
thanks

**Comment by endersstocker** [comment]
I posted the proposal to [the mailing list](https://groups.google.com/forum/#!topic/elixir-lang-core/scTj3Ifxrxc).

**Comment by endersstocker** [comment]
@josevalim I can understand why `Enum.none` would be omitted, given the ease of typing `not Enum.any?`, however there is some precedent in the language for providing complementary functions/macros. It might come as a surprise to a programmer to find certain complementary functions absent.

**Comment by SherSpock** [comment]
> @josevalim I can understand why `Enum.none` would be omitted, given the ease of typing `not Enum.any?`, however there is some precedent in the language for providing complementary functions/macros. It might come as a surprise to a programmer to find certain complementary functions absent.

On that note, was pairing with a friend this afternoon -- we were both surprised that `Enum.none?` isn't there and both consider it easier to read than `!Enum.any?`. 

I came straight here to see if adding `none?` had already been attempted.  

Though I am not a cognitive scientist or epistemologist, to the best of my understanding the human mind has more trouble reasoning about negation than presence. Hiding the implementation details of that negation behind a common and minimally ambiguous word seems much more approachable. 

Just from a language parsing perspective, you also don't have to first understand the expression and then calculate its opposite. The meaning is there in the expression itself. 

Also adding from following the [mailing list discussion](https://groups.google.com/forum/#!topic/elixir-lang-core/scTj3Ifxrxc) that it *is* available in [Crutches](https://github.com/mykewould/crutches)

**Pull request closed without merging**